src/v/raft/offset_translator.cc

```diff
@@ -394,6 +394,14 @@ ss::future<> offset_translator::sync_with_log(
     co_await std::move(reader).for_each_ref(
       log_consumer{*this}, model::no_timeout);
 
+    if (_highest_known_offset < log_offsets.dirty_offset) {
+        throw std::runtime_error{_logger.format(
```

Comment:
Perhaps this should be a vassert?  I think this assertion would go up to the controller_backend when it's starting a partition & cause that to retry continuously, which would be hard for a user to diagnose/understand.

Reply:
Well, maybe! I'm still not sure about what is the right policy for choosing between assertion/exception. It seems unfair to crash the whole server due to problems with one partition, but OTOH the user indeed has almost no way to determine what happened.